### PR TITLE
570 saved map searches

### LINKED ISSUE #570: Allow people to "save" searches of the map and get email updates for new posts in the search
- [x] Page in the user settings to view current saved searches
   - [x] You can click one to load it in the map
   - [x] You can also delete them
- [x] Add saved searches pop up to MapExplorer view
   - [x] Users can click button to save current search in the map
      - [x] Users can name their saved searches
      - [x] Users can view the filters in their saved search
   - [x] Users can view current saved searches
   - [x] Users can delete a saved search

depends on https://github.com/Hylozoic/hylo-node/issues/514

### PR DESCRIPTION
**Pull request by wushenshen**
Closes #570 

I added saved searches to the existing opt-in text on the notifications tab - chatted with @lorenjohnson about it, seems like a quick, acceptable way to handle opt-ins. 

A possible functional enhancement is to make each saved search accessible via a unique URL - currently the settings are all managed via state so the url is the same for all saved searches in each context (all, public, a community or network). 

Related: https://github.com/Hylozoic/hylo-node/pull/545

Screen Shot 2020-10-14 at 8 07 54 PM

Screen Shot 2020-10-14 at 8 08 38 PM

Screen Shot 2020-10-14 at 8 08 17 PM

Screen Shot 2020-10-14 at 8 08 22 PM

Screen Shot 2020-10-14 at 8 08 13 PM

### CODE REVIEW

**Review by lorenjohnson — COMMENTED** [review]
Looks great Shen Shen! Tidy implementation, easy to read and reason and a really valuable feature. 

I made a number of comments in the spirit of conversation and coherence for us as a team, only a one thing which I feel strongly about for the current work:

* Update existing tests to passing and dd some level of testing coverage for the feature.

Other than that please consider my question/remarks on the location of graphql, `let` vs `const` in a cpl places, and anything else that strikes you as worth pursuing in my comments.

**Comment by lorenjohnson** [comment]
In terms of UI I'd like to see the heart icon be white and go to green on hover and while the saved search drop down is active.

File: src/util/assets.js [hunk 1]
```diff
@@ -13,3 +13,5 @@ export const jollyAxolotl = '/jolly-axolotl.png'
 export const axolotlDigging = '/axolotl-digging.svg'
 export const axolotlPhone = '/axolotl-phone.png'
 export const mapIconAtlas = '/mapIconAtlas.png'
+export const heart = '/heart.svg'
```
[the remark this answers is not in the document]
Comment (tibetsprague) [reply src/util/assets.js:16]: maybe because the font wouldnt work for embedding in the webgl map?

File: src/routes/MapExplorer/MapExplorer.connector.js [hunk 2]
```diff
@@ -93,18 +98,31 @@ export function mapDispatchToProps (dispatch, props) {
...
     gotoMember: (memberId) => dispatch(push(personUrl(memberId, routeParams.slug, routeParams.networkSlug))),
     toggleDrawer: (hidden) => dispatch(push(addQuerystringToPath(baseUrl({ ...routeParams, view: 'map' }), { ...querystringParams, hideDrawer: hidden }))),
     storeFetchParams: param => opts => dispatch(storeFetchParams({ ...param, ...opts })),
-    storeClientFilterParams: params => dispatch(storeClientFilterParams(params))
+    storeClientFilterParams: params => dispatch(storeClientFilterParams(params)),
+    viewSavedSearch: (params, path, selectedSearch) => {
+      const { boundingBox, featureTypes, search, topics } = params
+      dispatch(fetchMembers({ ...params }))
+      dispatch(fetchPosts({ ...params }))
+      dispatch(fetchPublicCommunities({ ...params }))
```
Comment (tibetsprague) [inline src/routes/MapExplorer/MapExplorer.connector.js:114]: this smells a bit off to me, seems like viewing a search should just change the params and then the map should call all these on its own since/if the params have changed

File: src/routes/MapExplorer/MapExplorer.js [hunk 3]
```diff
@@ -6,19 +6,23 @@ import combine from '@turf/combine'
...
 import Icon from 'components/Icon'
 import Loading from 'components/Loading'
-import { FEATURE_TYPES } from './MapExplorer.store'
+import { FEATURE_TYPES, formatBoundingBox } from './MapExplorer.store'
 import Map from 'components/Map/Map'
 import MapDrawer from './MapDrawer'
+import SavedSearches from './SavedSearches'
 import { createIconLayerFromPostsAndMembers } from 'components/Map/layers/clusterLayer'
 import { createIconLayerFromCommunities } from 'components/Map/layers/iconLayer'
+import { heart } from 'util/assets'
```
Comment (tibetsprague) [inline src/routes/MapExplorer/MapExplorer.js:20]: i agree would be better to put this in the icon font, so lets go over that together

File: src/routes/MapExplorer/MapExplorer.js [hunk 4]
```diff
@@ -48,6 +52,8 @@ export default class MapExplorer extends React.Component {
       showFeatureFilters: false,
       totalLoadedBoundingBox: null,
       viewport: {
+        width: 800,
```
Comment (tibetsprague) [inline src/routes/MapExplorer/MapExplorer.js:55]: why is this needed?

File: src/routes/MapExplorer/MapExplorer.js [hunk 5]
```diff
@@ -90,6 +102,11 @@ export default class MapExplorer extends React.Component {
          !isEqual(prevProps.publicCommunities, publicCommunities))) {
       this.setState(this.updatedMapFeatures(this.state.currentBoundingBox))
     }
+
+    if (prevProps.selectedSearch !== this.props.selectedSearch) {
+      this.updateBoundingBoxQuery(this.props.selectedSearch.boundingBox)
```
Comment (tibetsprague) [inline src/routes/MapExplorer/MapExplorer.js:107]: so this is where it seems like the params would get updated which would automatically call the other update functions to load posts, etc.

File: src/routes/MapExplorer/MapExplorer.js [hunk 6]
```diff
@@ -252,13 +273,66 @@ export default class MapExplorer extends React.Component {
...
+  }
+
+  saveSearch = (name) => {
+    const { currentBoundingBox } = this.state
+    const { currentUser, filters, location, posts } = this.props
+    const { featureTypes, search: searchText, topics } = filters
+    const { pathname } = location
+
+    let context
+    if (isAllCommunitiesPath(pathname)) context = 'all'
```
Comment (tibetsprague) [inline src/routes/MapExplorer/MapExplorer.js:287]: yeah seems like context should be known and passed in from the .connector instead of determined everytime a search is saved here

File: src/routes/MapExplorer/MapExplorer.js [hunk 7]
```diff
@@ -252,13 +273,66 @@ export default class MapExplorer extends React.Component {
...
+    const topicIds = topics.map(t => t.id)
+
+    const boundingBox = [
+      { lat: currentBoundingBox[1], lng: currentBoundingBox[0] },
+      { lat: currentBoundingBox[3], lng: currentBoundingBox[2] }
+    ]
+
+    const attributes = { boundingBox, communitySlug, context, lastPostId, name, networkSlug, postTypes, searchText, topicIds, userId }
+
+    this.props.saveSearch(attributes).then(_ => this.props.fetchSavedSearches())
```
Comment (tibetsprague) [inline src/routes/MapExplorer/MapExplorer.js:316]: are you calling fetch to add it to the list? the reducer could also add it to the list, not a big deal but definitely more efficient than another server call

File: src/routes/UserSettings/UserSettings.connector.js [hunk 8]
```diff
@@ -63,21 +69,36 @@ export function mapStateToProps (state, props) {
...
+    unlinkAccount: (params) => dispatch(unlinkAccount(params)),
+    setConfirmBeforeClose: (params) => dispatch(setConfirmBeforeClose(params)),
+    updateMembershipSettings: (params) => dispatch(updateMembershipSettings(params)),
+    updateAllMemberships: (params) => dispatch(updateAllMemberships(params)),
+    registerStripeAccount: (params) => dispatch(registerStripeAccount(params)),
+    viewSavedSearch: (params, path, selectedSearch) => {
+      const { boundingBox, featureTypes, search, topics } = params
+      dispatch(fetchMembers({ ...params }))
+      dispatch(fetchPosts({ ...params }))
+      dispatch(fetchPublicCommunities({ ...params }))
```
Comment (tibetsprague) [inline src/routes/UserSettings/UserSettings.connector.js:95]: again this all feels weird, why do we need this in the UserSettings?

**Review by tibetsprague — CHANGES_REQUESTED** [review]

File: src/routes/MapExplorer/SavedSearches/SavedSearches.js [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+          <div styleName='searchBox'>
+            <input
+              type='text'
+              onChange={e => setName(e.target.value)}
+              placeholder='Name this view'
+              value={name}
+            />
+            <span styleName={`save ${canSave ? '' : 'disabled'}`} onClick={canSave ? () => saveSearch(name) : undefined}>Save</span>
+          </div>
+          <div styleName='filters'><img src={info} />&nbsp;&nbsp;{currentFilters(filters)} </div>
```
[the remark this answers is not in the document]
Comment (wushenshen) [reply src/routes/MapExplorer/SavedSearches/SavedSearches.js:46]: Agreed - this actually isn't needed, there's a left padding on the inner span. I took it out

File: src/routes/UserSettings/SavedSearchesTab/SavedSearchesTab.js [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+import React, { Component } from 'react'
+import './SavedSearchesTab.scss'
+import { formatParams, generateViewParams } from 'util/savedSearch'
+import { info } from 'util/assets'
+import Loading from 'components/Loading'
+import ReactTooltip from 'react-tooltip'
+
+const { array, func } = PropTypes
+
+export default class SavedSearchesTab extends Component {
```
[the remark this answers is not in the document]
Comment (wushenshen) [reply src/routes/UserSettings/SavedSearchesTab/SavedSearchesTab.js:11]: Yes - going to stick with the pattern of the other tabs for now

File: src/routes/MapExplorer/SavedSearches/SavedSearches.js [hunk 11]
```diff
@@ -0,0 +1,85 @@
+import React, { useState } from 'react'
+
+import { currentFilters, formatParams, formatParamPreview } from 'util/savedSearch'
+import Icon from 'components/Icon'
+import { info } from 'util/assets'
+// eslint-disable-next-line
+import styles from './SavedSearches.scss'
```
[the remark this answers is not in the document]
Comment (wushenshen) [reply src/routes/MapExplorer/SavedSearches/SavedSearches.js:7]: Ah I see thanks for the tip, I'm going to remove the name and the eslint comment

File: src/routes/MapExplorer/SavedSearches/SavedSearches.js [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+import { currentFilters, formatParams, formatParamPreview } from 'util/savedSearch'
+import Icon from 'components/Icon'
+import { info } from 'util/assets'
+// eslint-disable-next-line
+import styles from './SavedSearches.scss' 
+import ReactTooltip from 'react-tooltip'
+
+function SavedSearches (props) {
+  let {
```
[the remark this answers is not in the document]
Comment (wushenshen) [reply src/routes/MapExplorer/SavedSearches/SavedSearches.js:11]: Yep agreed, I've changed this to `const`

File: src/util/savedSearch.js [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+    public: `Public communities`,
+    all: `All communities`
+  }
+  return `${contextDetails[context]} • ${parsePostTypes(postTypes)}`
+}
+
+export function generateViewParams (search) {
+  const { context, community, network, postTypes, searchText, topics } = search
+
+  let { boundingBox } = search
```
[the remark this answers is not in the document]
Comment (wushenshen) [reply src/util/savedSearch.js:53]: Yes, moved it to L51 too. In a previous version I was mutating `boundingBox`, but no more.

File: src/routes/MapExplorer/MapExplorer.js [hunk 14]
```diff
@@ -252,13 +273,66 @@ export default class MapExplorer extends React.Component {
...
+    this.setState({ showSavedSearches: !this.state.showSavedSearches })
+  }
+
+  saveSearch = (name) => {
+    const { currentBoundingBox } = this.state
+    const { currentUser, filters, location, posts } = this.props
+    const { featureTypes, search: searchText, topics } = filters
+    const { pathname } = location
+
+    let context
```
[the remark this answers is not in the document]
Comment (wushenshen) [reply src/routes/MapExplorer/MapExplorer.js:286]: Ah yes - this can indeed be pulled out of `routeParams.context`, and @tibetsprague made some changes to `MapExplorer.connector` that maps the context to subject. I'll remove this

File: src/routes/MapExplorer/MapExplorer.js (hunk 6 above)
Comment (wushenshen) [reply src/routes/MapExplorer/MapExplorer.js:287]: Yes - @lorenjohnson also commented on this above, I see that you'd mapped c to community and n to network as the subject, so I'm passing that in as the context.

File: src/routes/MapExplorer/MapExplorer.js (hunk 4 above)
Comment (wushenshen) [reply src/routes/MapExplorer/MapExplorer.js:55]: This allows the map to be zoomed in when loading a saved search from the user settings section. Otherwise, it isn't zoomed in at all.

File: src/graphql/mutations/CreateSavedSearchMutation.graphql [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+    context: $context,
+    lastPostId: $lastPostId,
+    name: $name,
+    networkSlug: $networkSlug,
+    postTypes: $postTypes,
+    searchText: $searchText,
+    topicIds: $topicIds,
+    userId: $userId
+  }) {
+    id
```
[the remark this answers is not in the document]
Comment (wushenshen) [reply src/graphql/mutations/CreateSavedSearchMutation.graphql:25]: @lorenjohnson I meant to return the actual savedSearch record in the backend, but it doesn't look like that's happening.  I'm only getting the ID.

File: src/routes/MapExplorer/MapExplorer.js (hunk 7 above)
Comment (wushenshen) [reply src/routes/MapExplorer/MapExplorer.js:316]: This is related to my comment above, I'm not getting the full saved search record after creating a saved search

File: src/routes/MapExplorer/SavedSearches/SavedSearches.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (wushenshen) [reply src/routes/MapExplorer/SavedSearches/SavedSearches.js:46]: Oops correction - was looking at the wrong span, I did end up adding the right margin.

File: src/routes/UserSettings/UserSettings.connector.js (hunk 8 above)
Comment (wushenshen) [reply src/routes/UserSettings/UserSettings.connector.js]: Was to provide the necessary updates to the store... but I've replaced that with [this](https://github.com/Hylozoic/hylo-evo/pull/663/commits/d76cbcc57adebd6875c90ba931d684890966c35d#diff-fbe00c3522dc75540d6eedef7bce90e7df0a0c88efd8db27db585ba75f19223bR111)

File: src/routes/MapExplorer/MapExplorer.js (hunk 5 above)
Comment (wushenshen) [reply src/routes/MapExplorer/MapExplorer.js]: Good point, I've updated that to use `this.updateSavedSearch` which is also called in `componentDidMount`

File: src/util/assets.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (wushenshen) [reply src/util/assets.js]: I've removed the heart/info svgs from this file and added them to the font instead!

File: src/graphql/mutations/CreateSavedSearchMutation.graphql (hunk 15 above)
[the remark this answers is not in the document]
Comment (wushenshen) [reply src/graphql/mutations/CreateSavedSearchMutation.graphql]: @lorenjohnson I changed it to return `savedSearch { id, ... }` as you suggested

File: src/graphql/mutations/CreateSavedSearchMutation.graphql [hunk 16]
```diff
@@ -0,0 +1,27 @@
+mutation CreateSavedSearchMutation(
```
[the remark this answers is not in the document]
Comment (wushenshen) [reply src/graphql/mutations/CreateSavedSearchMutation.graphql:1]: @lorenjohnson No particular reason - I think because the query was larger than the others

File: src/routes/MapExplorer/MapExplorer.js (hunk 7 above)
Comment (wushenshen) [reply src/routes/MapExplorer/MapExplorer.js]: @tibetsprague I'm using the reducer to add it to the list now, see this [latest commit](https://github.com/Hylozoic/hylo-evo/pull/663/commits/a740eb8d09dfd825db56e012a96fb10209c1a30f)

File: src/css/global/icons.scss [hunk 17]
```diff
@@ -199,9 +196,18 @@
...
 }
-.icon-Copy:before {
-  content: "\e93d";
+.icon-Edit:before {
+  content: "\1f323";
+}
+.icon-Heart:before {
+  content: "\e93e";
 }
+.icon-Info:before {
```
Comment (wushenshen) [inline src/css/global/icons.scss:211]: I'm not sure why these still ended up getting reordered, I used the latest icons svg from the `dev` branch. Thinking I'll leave them, if that's okay with y'all.

Comment (lorenjohnson) [reply src/css/global/icons.scss:211]: It looks fine to me. The order they appear in the CSS doesn't matter and the content keys for the existing icons don't seem to have shifted. 👍

**Review by lorenjohnson — APPROVED** [review]
👍  LGTM

**Review by tibetsprague — APPROVED** [review]

**Comment by tibetsprague** [comment]
@wushenshen this is almost ready to go but do fix the tests so at least they pass!

**Comment by wushenshen** [comment]
@tibetsprague ok - tests are passing now!
